freqtrade/templates/ExamplePredictionModel.py

```diff
@@ -0,0 +1,159 @@
+            self.dh.data["avg_mean_dist"] = self.dh.compute_distances()
+
+        logger.info("length of train data %s", len(data_dictionary["train_features"]))
+
+        model = self.fit(data_dictionary)
+
+        logger.info(f'--------------------done training {metadata["pair"]}--------------------')
+
+        return model
+
+    def fit(self, data_dictionary: Dict) -> Any:
+        """
+        Most regressors use the same function names and arguments e.g. user
+        can drop in LGBMRegressor in place of CatBoostRegressor and all data
+        management will be properly handled by Freqai.
+        :params:
+        :data_dictionary: the dictionary constructed by DataHandler to hold
+        all the training and test data/labels.
+        """
+
+        train_data = Pool(
+            data=data_dictionary["train_features"],
+            label=data_dictionary["train_labels"],
+            weight=data_dictionary["train_weights"],
+        )
+
+        test_data = Pool(
+            data=data_dictionary["test_features"],
+            label=data_dictionary["test_labels"],
+            weight=data_dictionary["test_weights"],
+        )
+
+        model = CatBoostRegressor(
+            verbose=100, early_stopping_rounds=400, **self.model_training_parameters
+        )
+        model.fit(X=train_data, eval_set=test_data)
+
+        return model
+
+    def predict(self, unfiltered_dataframe: DataFrame, metadata: dict) -> Tuple[DataFrame,
+                                                                                DataFrame]:
+        """
+        Filter the prediction features data and predict with it.
+        :param: unfiltered_dataframe: Full dataframe for the current backtest period.
+        :return:
+        :predictions: np.array of predictions
+        :do_predict: np.array of 1s and 0s to indicate places where freqai needed to remove
+        data (NaNs) or felt uncertain about data (PCA and DI index)
+        """
+
+        # logger.info("--------------------Starting prediction--------------------")
+
+        original_feature_list = self.dh.build_feature_list(self.config, metadata)
+        filtered_dataframe, _ = self.dh.filter_features(
+            unfiltered_dataframe, original_feature_list, training_filter=False
+        )
+        filtered_dataframe = self.dh.standardize_data_from_metadata(filtered_dataframe)
+        self.dh.data_dictionary["prediction_features"] = filtered_dataframe
+
+        # optional additional data cleaning
+        if self.feature_parameters["principal_component_analysis"]:
+            pca_components = self.dh.pca.transform(filtered_dataframe)
+            self.dh.data_dictionary["prediction_features"] = pd.DataFrame(
+                data=pca_components,
+                columns=["PC" + str(i) for i in range(0, self.dh.data["n_kept_components"])],
+                index=filtered_dataframe.index,
+            )
+
+        if self.feature_parameters["remove_outliers"]:
```

Comment:
This is becomming danger zone.
as this is now part of the "user-defined" model - i'm not sure if having it in the config makes sense (it could also be defined in this file).

The problem is - we only have / check X values for correctness (only the ones we know / use in this example).
People might however have 20 different desires for this - and placing all in the config might become problematic.

if we assume this to be "automatic" things - then the default shouldn't have it in the example file - but someplace in the backend.
if we assume the user modifies it (also what it does) - then the config is not the right place. 

It's difficult to control what people put within the `remove_outliers` IF statement (you can't, really)- so you can no longer rely on the fact that it does what the description says.

Reply:
Yeah, I agree with what you are saying here. Let me look into transitioning these userconfigurable options over to the hyperparameters type setup so that they can put them in the strategy directly. 

Maybe ALL freqai stuff should go in there...I need to spend time learning how it works and then I will come back to with a proposed solution.

Reply:
maybe - maybe not - that's ... difficult to determine.

i think you'll want at least some setup parameters in the config - so freqtrade can determine when to retrain.